---
title: "Introduction to MethICA"
author: "Lea Meunier"
date: "5/31/2020"
output:
  html_document: default
  pdf_document: default
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Abstract

<p>DNA methylation changes are widespread in human cancers, but the underlying molecular mechanisms remain incompletely understood. We developed an innovative statistical framework, MethICA, leveraging independent component analysis to identify sources of DNA methylation changes in tumors. The package includes a function that uses independent component analysis to extract epigenetic signatures from methylation data, as well as functions to calculate associations with sample annotations and CpG characteristics. The package also provides representations that facilitate the interpretation of methylation components. This document, paired with the “MethICA_examples_script.R” demo script, outlines the typical workflow for analyzing methylation signatures in a cancer series with MethICA.</p>

# Package

Report issues at <https://github.com/FunGeST/MethICA>.


\newpage

# Introduction


# Installation Instructions

The latest version of the package can be installed from the FunGeST GitHub repository using devtools:
```{r install, eval = FALSE, echo = TRUE}
install.packages("devtools")
library(devtools)
devtools::install_github("FunGeST/MethICA")

```
# Dependencies

<p>The R packages stringr, fastICA, cowplot, ggplot2, RColorBrewer, plotrix and broom are required to perform MethICA analysis</p>

# Input data

Input files are necessary to perform the core MethICA analyses:
<ul>
<li>bval: methylation levels for each CpG or region (rows) in each sample (columns)
<li>CpG annotation: CpG table annotated with various (epi)genomic features 
<li>sample annotation: relevant sample annotations to interpret the components
</ul>

<p>Please check the README file for detailed description of input files. Examples are also provided with the package.</p>



## Load methylation data and annotations

Once installed, load the package and you’re ready to go! 
```{r chargepackage, eval = FALSE, echo = TRUE}
> # Load MethICA package
> library(MethICA)
```

Define data and output directories.
```{r chargedata, eval = FALSE, echo = TRUE}
> # define input directory containing example dataset> 
> datadir <- "~/Documents/MethICA/RUNNING_MethICA_EXAMPLE/LICAFR"
>
> # define output directory> 
> output.directory <- "~/Results/"> 
> if(!file.exists()){
>  dir.create(output.directory)
> }
```

We provide example datasets from our hepatocellular carcinoma study that can be loaded as follows:
```{r chargeLICAFR, eval = FALSE, echo = TRUE}

> # carge example dataset> 
> bval = load.RData(file.path(datadir, 'RUNNING_MethICA_exemple/LICAFR/bval.Rdata'))
> annot = load.RData(file.path(datadir, 'RUNNING_MethICA_exemple/LICAFR/LICAFR_annot.RData'))
```

Select the most variant CpG sites (based on standard deviation) for the analysis.
```{r selectMostVar, eval = FALSE, echo = TRUE}
> # Select most variant CpG sites 
> NmostVar = 200000
> mysd <- apply(bval,1,sd)
> sel <- order(mysd,decreasing=T)[1:NmostVar]
> # Reduce bval and CpG_feature matrix
> bval <- bval[sel,];dim(bval)
> CpG_feature <- CpG_feature[rownames(bval),]
```

## Prepare CpG annotation table

MethICA uses various (epi)genomic annotations of CpG sites to interpret methylation components. Make sure you use correct annotations for the tissue under study. For example, the CpG_feature.Rdata file included in the package corresponds the CpG annotation table of liver tissue used in our hepatocellular carcinoma study. We provide the chromatin.feature function to annotate your own CpG table. It requires different inputs for each (epi)genomic feature that can be obtained from various sources:
<ul>
<li>file_CGI : CpG island-based features (Island, Shore, Shelf, out of cgi) from UCSC
<li>file_genes : gene-based features (body, TSS500) from GENCODE <https://www.gencodegenes.org/human/release_34lift37.html>
<li>file_chrom_state : chromatin states defined from various histone marks by the ROADMAP epigenomics project <https://egg2.wustl.edu/roadmap/web_portal/chr_state_learning.html#exp_18state>
<li>file_CpG_context : methylation domains (HMD/PMD/LMR/UMR) defined from WGBS data <https://www.ncbi.nlm.nih.gov/geo/download/?acc=GSE113405&format=file&file=GSE113405%5FLIV%5FADLT%2EMethylSeekR%2Esegments%2Ebed%2Egz>
<li>file_replication : replication timing deciles obtained from Repli-Seq data availbale on the ENCODE project data portal. Here we used Repli-Seq from HepG2 cell line accessible under GEO accession number GSM923446 <https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSM923446> 
</ul>

```{r CpG_feature, eval = FALSE, echo = TRUE}
> # Annotate CpG table with various epigenomic features
> #CpG_table = load.RData('~/Google Drive/MethICA/data/CPG_feature_Illumina.RData')
> CpG_feature = chromatin.feature(CpG_table = CpG_table, file_CpG_context = "MethICA/data/GSE113405_LIV_ADLT.MethylSeekR.segments.bed", name_col_CpG_context = "CpG_context", file_chrom_state = 'MethICA/data/cst18_liver.RData', name_col_chrom_state = c("state", "active"), file_CGI = "MethICA/data/CGI-based_features_hg19.txt", name_col_CGI ="cgi_feature", file_genes = "MethICA/data/Gene-based_features_hg19.txt", name_col_genes = c("gene_name", "gene_feature"), file_replication ="MethICA/data/HepG2_replication_domains.RData", name_col_replication = "decile", add_seq_info = TRUE, save = TRUE, output.directory = output.directory)
```

# Extract methylation components with ICA

The mc.extract function performs independent component analysis (ICA) and extracts methylation components from the methylation matrix.
<ul>
<li> input: bval methylation matrix 
<li> outputs: MC_object with two matrices giving the contribution of CpGs and samples to each component, and one vector giving components stability. If compute_stability = TRUE (recommended), mc.extract performs n iterations of ICA, computes stability and selects the most stable iteration,. If compute_stability = FALSE, mc.extract performs a single iteration of ICA and returns NA in stability vector
</ul>

```{r ICA, eval = FALSE, echo = TRUE}
> MC_object <- mc.extract(bval, nb_comp = 20, compute_stability = TRUE, nb_iteration = 100, output.directory = output.directory, save = TRUE)
```

<p>Each methylation component (MC) is characterized by an activation pattern across CpG sites and across samples. To interpret their biological meaning, we first select the most contributing CpGs and samples for each MC.</p>

<p>The mc.active.CpG function identifies CpGs with a contribution greater than a defined threshold (method="threshold", recommended) or extracts a defined number of most contributing CpGs (method="number").</p>

<p>The mc.active.sample function identifies the most contributing samples (method="absolute") or those showing the greatest deviation from a set of reference samples (method="reference").</p>


```{r contributing, eval = FALSE, echo = TRUE}
> # Extract the most contributing CpG sites for each MC
> MC_contrib_CpG <- mc.active.CpG(MC_object, method = "threshold")
>
> # Extract the most differentially methylated samples for each MC
> # Extract the most contributing samples for each MC based on absolute value of contribution 
> MC_active_sample = mc.activ.sample(MC_object, method = c("absolute", "reference")[1],bval = bval , MC_contrib_CpG = > MC_contrib_CpG, number = round(nrow(MC_object$Sample_contrib)*0.1), output.directory = output.directory)
> # Extract the most contributing samples for each MC based on differential methylation level with reference sample (here normal samples)
> MC_active_sample = mc.activ.sample(MC_object, method = c("absolute", "reference")[2],bval = bval , MC_contrib_CpG = > MC_contrib_CpG, number = round(nrow(MC_object$Sample_contrib)*0.1), ref = grep("N", colnames(bval), value = TRUE), output.directory = output.directory)
```

# Represent methylation changes

<p>We then use the mc.change function to identify the major methylation changes associated with each component. This function plots the average methylation of the most contributing CpGs in the most contributing samples versus reference samples. Examples below represent components associated mostly with hypermethylation, hypomethylation or both. If highly contributing samples include samples with high positive and negative contributions, two distinct graphs are produced. </p>

```{r MC_change, eval = FALSE, echo = TRUE}
> #Represent methylation changes in most contributing tumors vs. normal samples
> mc.change(MC_object, MC_active_sample, MC_contrib_CpG, bval, ref = grep("N", colnames(bval), value = TRUE), output.directory = output.directory)
```

Examples of outputs: 
![methylation change](./meth_change.png){width=100% height=100%}

# Explore epigenomic characteristics of the most contributing CpGs

To better understand the components, we then explore the characteristics of their most contributing CpGs. The enrich.CpG.feature function computes enrichment scores of CpGs across epigenomic features from the CpG_feature table and generates various visual outputs. The example below shows a hypermethylation component affecting preferentially CpG sites located in CpG islands near transcription start sites, with bivalent chromatin state.

```{r enrich_CpG, eval = FALSE, echo = TRUE}
> # Association of MCw with (epi)genomic characteristics
> enrich.CpG.feature(MC_object, MC_contrib_CpG, output.directory = output.directory, CpG_feature = CpG_feature)
```

Example of outputs for MC7 component: 
![methylation change](./CpG_feature.png){width=80% height=80%}

<p> In the Zhou et al. paper (Zhou, W., Dinh, H.Q., Ramjan, Z., Weisenberger, D.J., Nicolet, C.M., Shen, H., Laird, P.W., and Berman, B.P. (2018). DNA methylation loss in late-replicating domains is linked to mitotic cell division. Nature Genetics 50, 591–602.) CpG are split in 48 categories depending of the methylation domain, number of flanking CpG and adjacent nucleotide, and the team observe different methylation modification during time. In MethICA with add this CpG information, performe an enrichment analysis and representation to have in on figure, all the enrichment results.</p>

```{r enrich_CpG_Zhou, eval = FALSE, echo = TRUE}
# create table with catagorie
CpG_feature$nb_flanking_CpG_reccod = CpG_feature$nb_flanking_CpG
CpG_feature$nb_flanking_CpG_reccod[which(CpG_feature $nb_flanking_CpG_reccod>3)] = 3

enrich_context = matrix(NA, nrow = nrow(CpG_feature), ncol = 4*4*3)
rownames(enrich_context) = rownames(CpG_feature)
colnames(enrich_context) = paste(rep(c("HMD", "PMD", "LMR", "UMR"), each = 12), rep(rep(3:0, each = 3), length = 48), rep(c("SCGS", "SCGW", "WCGW"), length = 48), sep = "_")

i=1
for(domain_type in c("HMD", "PMD", "LMR", "UMR")){
	for(nb_flanking_CpG in c(3, 2, 1, 0)){
		for(context in c("SCGS", "SCGW", "WCGW")){
			enrich_context[which(CpG_feature$CpG_context == domain_type & CpG_feature$nb_flanking_CpG_reccod == nb_flanking_CpG & CpG_feature$context == context),i] = "yes"
			i=i+1
		}
	}
}
enrich_context = rbind(enrich_context, rep("yes", ncol(enrich_context)))
table_enrich_context = apply(enrich_context, 2, table)[colnames(enrich_context)]-1

# compute enrichment in each component
nbComp = length(MC_active_sample)

matrice_enrich = matrix(NA, ncol = nbComp, nrow = 48)
rownames(matrice_enrich) = names(table_enrich_context)
colnames(matrice_enrich) = paste0("MC", 1: nbComp)

for(i in 1:nbComp){
	context = enrich_context[MC_contrib_CpG[[i]],]
	context = rbind(context, rep("yes", ncol(context)))
		
	table_context = apply(context, 2, table)[colnames(enrich_context)]-1
	matrice_enrich[,i] = (table_context / sum(table_context)) / (table_enrich_context / sum(table_enrich_context))

}

# scale for representation
matrice_enrich = matrice_enrich-1 
matrice_enrich[which(matrice_enrich>5)] = 5
matrice_enrich[which(matrice_enrich <0)] = 0

cst_color <- colorRampPalette(c("white","white", "grey40"))(100 + 1)

pdf(file.path(output.directory, "CpG_context_Zhou.pdf"),width=8,height=8, bg = "transparent")
par(oma = c(0,0,0,0), xpd=TRUE, col = "white", mar=c(0,0,0,0), bg = "transparent")	
corrplot(matrice_enrich, col = cst_color, is.corr = FALSE, tl.col = "black", tl.cex = 1.1, mar=c(0,0,0,0), bg = "transparent",cl.lim = c(0,6))
dev.off()
```

<p> Example of outputs for 48 CpG context in the 20 components: </p>
![methylation change](./CpG_context_Zhou.png){width=40% height=40%}


# Association with sample annotations

<p>The final step is to analyze the characteristics of samples most strongly contributing to each component. The mc.annot function first performs univariate linear regressions to identify annotations associated with each MC. Significant annotations are then included in multivariate analyses to determine the strongest determinants of each MC.</p>

```{r assos_annot, eval = FALSE, echo = TRUE}
> # Association of MCs with clinical and molecular features
> Samples_association = mc.annot(MC_object, annot = annot , save = TRUE, output.directory = output.directory, seuil_multi = 0.001)
```

Examples of representations for associations between annotations and components (left = univariate, rigth = multivariate): 
![methylation change](./corrplot.png){width=90% height=90%}
<p>p-value circle/color legend (see echelle_log on the MethICA_example_script.R)<p>
<ul>
<li>1 = 1
<li>0.1 = 17
<li>0.05 = 22
<li>0.01 = 33
<li>1.0 10-4 = 65
<li>1.0 10-6 = 96
<li>1.0 10-8 = 128
<li>0 = 160
</ul>

![](./MC13_vs_annot.png){width=30% height=30%}
